station-boards-injector/src/main/java/stream/DelayedTrainsListenerVerticle.java

```diff
@@ -0,0 +1,109 @@
+package stream;
+
+import datamodel.Station;
+import datamodel.Stop;
+import datamodel.Train;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.json.JsonObject;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.client.hotrod.marshall.ProtoStreamMarshaller;
+import org.infinispan.commons.util.Util;
+import org.infinispan.protostream.FileDescriptorSource;
+import org.infinispan.protostream.SerializationContext;
+import org.infinispan.query.api.continuous.ContinuousQuery;
+import org.infinispan.query.api.continuous.ContinuousQueryListener;
+import org.infinispan.query.dsl.Query;
+import org.infinispan.query.dsl.QueryFactory;
+import org.infinispan.query.remote.client.ProtobufMetadataManagerConstants;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.infinispan.query.remote.client.ProtobufMetadataManagerConstants.PROTOBUF_METADATA_CACHE_NAME;
+import static stream.Util.mkRemoteCacheManager;
+
+public class DelayedTrainsListenerVerticle extends AbstractVerticle {
+
+  private RemoteCacheManager client;
+  private RemoteCache<String, Stop> stopsCache;
+  private ContinuousQuery<String, Stop> delayedTrainsCQ;
+
+  @Override
+  public void start() throws Exception {
+    client = mkRemoteCacheManager();
+    client.getCache("default").clear();
+
+    addProtoDescriptorToServer(client.getCache(PROTOBUF_METADATA_CACHE_NAME));
+
+    addDelayedTrainsListener();
+  }
+
+  @Override
+  public void stop() throws Exception {
+    if (delayedTrainsCQ != null)
+      delayedTrainsCQ.removeAllListeners();
+
+    if (stopsCache != null)
+      stopsCache.clear();
+
+    if (client != null)
+      client.stop();
+  }
+
+  private void addDelayedTrainsListener() {
+    stopsCache = client.getCache("default");
+    QueryFactory qf = Search.getQueryFactory(stopsCache);
+
+    Query query = qf.from(Stop.class)
+      .having("delayMin").gt(0L)
+      .build();
+
+    ContinuousQueryListener<String, Stop> listener =
+      new ContinuousQueryListener<String, Stop>() {
+        @Override
+        public void resultJoining(String id, Stop stop) {
+          vertx.runOnContext(x -> {
```

Comment:
You don't need this. Publish and send use a special context when not invoked on a Vert.x thread.

Reply:
And that's probably the case here...? The continuous query listener is invoked by an inner Java Hot Rod client thread.

Reply:
Yes, you can call `EventBus#send` and  `EventBus#publish` from any thread. Vert.x will use a special context (created only once when the event bus is created). Whereas `vertx.runOnContext` will keep creating new contexts (as not context is associated with the hotrod client thread).

Reply:
I had not forgotten about this. I've just [fixed it](https://github.com/infinispan-demos/streaming-data-workshop/commit/e36da2b308bf4497fcdb6fc62c5343489d40ed82).